Sources/NIO/Codec.swift

```diff
@@ -790,3 +790,284 @@ extension MessageToByteHandler {
+    }
+
+    private var decoder: Decoder
+    private let maximumBufferSize: Int?
+    internal private(set) var _buffer: ByteBuffer?
+
+    /// Initialize a `NIOSingleStepByteToMessageProcessor`.
+    ///
+    /// - parameters:
+    ///     - decoder: The `NIOSingleStepByteToMessageDecoder` to decode the bytes into message.
+    ///     - maximumBufferSize: The maximum number of bytes to aggregate in-memory.
+    public init(_ decoder: Decoder, maximumBufferSize: Int? = nil) {
+        self.decoder = decoder
+        self.maximumBufferSize = maximumBufferSize
+    }
+
+    private func append(_ buffer: ByteBuffer) {
+        if self._buffer == nil || self._buffer!.readableBytes == 0 {
+            self._buffer = buffer
+        } else {
+            var buffer = buffer
+            self._buffer!.writeBuffer(&buffer)
+        }
+    }
+
+    private func withNonCoWBuffer(_ body: (inout ByteBuffer) throws -> Decoder.InboundOut?) throws -> Decoder.InboundOut? {
+        guard var buffer = self._buffer else {
+            return nil
+        }
+
+        if buffer.readableBytes == 0 {
+            return nil
+        }
+
+        self._buffer = nil  // To avoid CoW
+        let result = try body(&buffer)
+        self._buffer = buffer
+        return result
+    }
+
+    private func decodeLoop(decodeMode: DecodeMode, seenEOF: Bool = false, _ messageReceiver: (Decoder.InboundOut) throws -> Void) throws {
+        // we want to call decodeLast once with an empty buffer if we have nothing
+        if decodeMode == .last && (self._buffer == nil || self._buffer!.readableBytes == 0) {
+            var emptyBuffer = self._buffer == nil ? ByteBufferAllocator().buffer(capacity: 0) : self._buffer!
+            if let message = try self.decoder.decodeLast(buffer: &emptyBuffer, seenEOF: seenEOF) {
+                try messageReceiver(message)
+            }
+            return
+        }
+
+        // buffer can only be nil if we're called from finishProcessing which is handled above
+        assert(self._buffer != nil)
+
+        func decodeOnce(buffer: inout ByteBuffer) throws -> Decoder.InboundOut? {
+            defer {
+                if buffer.readableBytes > 0 {
+                    if self.decoder.shouldReclaimBytes(buffer: buffer) {
+                        buffer.discardReadBytes()
+                    }
+                }
+            }
+            if decodeMode == .normal {
+                return try self.decoder.decode(buffer: &buffer)
+            } else {
+                return try self.decoder.decodeLast(buffer: &buffer, seenEOF: seenEOF)
+            }
+        }
+
+        while let message = try self.withNonCoWBuffer(decodeOnce) {
+            try messageReceiver(message)
+        }
+
+        if let maximumBufferSize = self.maximumBufferSize, self._buffer!.readableBytes > maximumBufferSize {
+            throw ByteToMessageDecoderError.PayloadTooLargeError()
+        }
+
+        if self._buffer!.readableBytes == 0 {
```

Comment:
Can you place a comment above these two blocks explaining why these two force unwraps are safe? (This is an SSWG best practice)